apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,285 @@
+   * @return {boolean} true if property access should be blocked.
+   */
+  shouldBlockCustomMarshalling_(name, obj, nativeParent) {
+    if (-1 !== this.customMarshaler.blockedProperties.indexOf(name)) {
+      return true;
+    }
+    var value = obj.isCustomMarshal ? obj.data[name] : nativeParent[name];
+    if (value instanceof Node || value instanceof Window) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Wrapper to Interpreter's getProperty (extended for custom marshaling)
+   *
+   * Fetch a property value from a data object.
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @return {!Object} Property value (may be UNDEFINED).
+   * @override
+   */
+  getProperty(obj, name) {
+    name = name.toString();
+    var customMarshalValue;
+    if (obj.isCustomMarshal) {
+      if (this.shouldBlockCustomMarshalling_(name, obj)) {
+        return this.UNDEFINED;
+      } else {
+        customMarshalValue = obj.data[name];
+      }
+    } else {
+      const nativeParent = this.getNativeParent_(obj, name);
+      if (nativeParent) {
+        customMarshalValue = nativeParent[name];
+      } else {
+        return super.getProperty(obj, name);
+      }
+    }
+    var type = typeof customMarshalValue;
+    if (type === 'number' || type === 'boolean' || type === 'string' ||
+        type === 'undefined' || customMarshalValue === null) {
+      return this.createPrimitive(customMarshalValue);
+    } else {
+      return codegen.marshalNativeToInterpreter(this, customMarshalValue, obj.data);
+    }
+  }
+
+  /**
+   * Returns the native parent of a custom marshaled object, if one exists.
+   * Specific cases that are accounted for are detailed in the conditional below.
+   *
+   */
+  getNativeParent_(obj, name) {
+    var hasProperty = false;
+    if (obj && !obj.isPrimitive) {
+      hasProperty = super.hasProperty(obj, name);
+    }
+    var nativeParent;
+    if (
+      /**
+       * 1. the values of custom marshaled globals can be overridden into interpreter values
+       *    on the global scope. If a variable with the same name has been declared on the
+       *    global scope, then don't try to look up the native parent.
+       */
+      !hasProperty &&
+      /**
+       * 2. custom marshaled objects can only be mounted on the global scope. Therefore
+       *    only look up a native parent if we are asking for a property on the global scope.
+       */
+      obj === this.globalScope &&
+      /**
+       * 3. Assuming the above conditions pass, lookup the native parent among the list
+       * of global properties specified in the custom marshaler's configuration.
+       */
+      !!(nativeParent = this.customMarshaler.globalProperties[name]) &&
+      /**
+       * 4. If the property being looked up has been explicitly blocked from custom
+       * marshalling, then don't return the native parent.
+       */
+      !this.shouldBlockCustomMarshalling_(name, obj, nativeParent)
```

Comment:
:rofl: Great commenting here.  Scary to even have a conditional this complicated and subtle, but great explanation.